neqo-transport/src/send_stream.rs

```diff
@@ -1063,61 +1115,290 @@ impl ::std::fmt::Display for SendStream {
+    //
+    // Two options are to walk the 'map' first, ignoring WebTransport
+    // streams, then process the unordered and ordered WebTransport
+    // streams.  The second is to have a sorted Vec for unfair streams (and
+    // use a normal iterator for that), and then chain the iterators for
+    // the unordered and ordered WebTranport streams.  The first works very
+    // well for H3, and for WebTransport nodes are visited twice on every
+    // processing loop.  The second adds insertion and removal costs, but
+    // avoids a CPU penalty for WebTransport streams.  For now we'll do #1.
+    //
+    // So we use a sorted Vec<> for the regular streams (that's usually all of
+    // them), and then a BTreeMap of an entry for each SendOrder value, and
+    // for each of those entries a Vec of the stream_ids at that
+    // sendorder.  In most cases (such as stream-per-frame), there will be
+    // a single stream at a given sendorder.
+
+    // These both store stream_ids, which need to be looked up in 'map'.
+    // This avoids the complexity of trying to hold references to the
+    // Streams which are owned by the IndexMap.
+    sendordered: BTreeMap<SendOrder, OrderGroup>,
+    regular: OrderGroup, // streams with no SendOrder set, sorted in stream_id order
+}
 
 impl SendStreams {
     pub fn get(&self, id: StreamId) -> Res<&SendStream> {
-        self.0.get(&id).ok_or(Error::InvalidStreamId)
+        self.map.get(&id).ok_or(Error::InvalidStreamId)
     }
 
     pub fn get_mut(&mut self, id: StreamId) -> Res<&mut SendStream> {
-        self.0.get_mut(&id).ok_or(Error::InvalidStreamId)
+        self.map.get_mut(&id).ok_or(Error::InvalidStreamId)
     }
 
     pub fn exists(&self, id: StreamId) -> bool {
-        self.0.contains_key(&id)
+        self.map.contains_key(&id)
     }
 
     pub fn insert(&mut self, id: StreamId, stream: SendStream) {
-        self.0.insert(id, stream);
+        self.map.insert(id, stream);
+    }
+
+    pub fn set_sendorder(&mut self, stream_id: StreamId, sendorder: Option<SendOrder>) -> Res<()> {
+	// check using non-mutable first so we can call set_fairness
+        if let Some(stream) = self.map.get(&stream_id) {
+	    // All sendordered streams are by definition fair since they're all WebTransport streams
+	    if !stream.fair {
+		self.set_fairness(stream_id, true)?;
+	    }
+	}
+        if let Some(stream) = self.map.get_mut(&stream_id) {
+            // don't grab stream here; causes borrow errors
+            let old_sendorder = stream.sendorder();
+            if old_sendorder != sendorder {
+                // we have to remove it from the list it was in, and reinsert it with the new
+                // sendorder key
+                let mut group = if let Some(old) = old_sendorder {
+                    self.sendordered.get_mut(&old).unwrap()
+                } else {
+                    &mut self.regular
+                };
```

Comment:
This might benefit from a function: `fn group_mut(&mut self, so: Option<SendOrder>) -> &mut Group { ... }`